### PR TITLE
add functions for trace, cofactor(minor), determinant, print

### PR DESCRIPTION
**Pull request by JakubPiotrowicz**

### CODE REVIEW

File: utils.h
Comment (dominikstrak2001) [inline utils.h:1]: I think it'd be cleaner if those were stattic methods on a Matrix class, ideally in `matrix.h` file

File: matrix.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                outputStream << matrix[i][j];
+                if (j != size.second - 1) {
+                    outputStream << " ";
+                }
+            }
+            outputStream << "\n";
+        }
+    }
+
+    bool vectorForMatrixValidator(std::vector<std::vector<T>> mat){
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
    bool vectorForMatrixValidator(std::vector<std::vector<T>> &mat){
```

File: matrix.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        if(mat.size() == 0){
+            return false;
+        }
+        int rowSize = mat[0].size();
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        int rowSize = mat[0].size();
        if (rowSize == 0) {
            return false;
        }
```

Comment (dominikstrak2001) [reply matrix.h]: I'd rename `rowSize` to `numberOfColumns` or `numColumns`

File: matrix.h [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        return true;
+    }
+
+public:
+    Matrix(std::vector<std::vector<T>> mat){
+        size = {mat.size(), mat[0].size()};
+        isSquare = size.first == size.second;
+        matrix = mat;
+
+        assert(vectorForMatrixValidator(mat));
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        assert(vectorForMatrixValidator(mat));

        size = {mat.size(), mat[0].size()};
        isSquare = size.first == size.second;
        matrix = mat;
```

Comment (dominikstrak2001) [reply matrix.h]: (moved assert to be the first statement)

File: matrix.h [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            case Solution:
+                printForSolutionTo(outputStream);
+                break;
+        }
+    }
+
+    static Matrix readMatrix(std::istream &inputStream) {
+        std::pair<size_t, size_t> size;
+        inputStream >> size.first >> size.second;
+        std::vector<std::vector<long long>> matrix(size.first, std::vector<long long>(size.second));
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        std::vector<std::vector<T>> matrix(size.first, std::vector<T>(size.second));
```

File: matrix.h [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        for (int i = 0; i < size.first; i++) {
+            for (int j = 0; j < size.second; j++) {
+                inputStream >> matrix[i][j];
+            }
+        }
+        return Matrix(matrix);
+    }
+
+    static Matrix constructIdentityMatrix(int size) {
+        std::vector<std::vector<long long>> matrix(size, std::vector<long long>(size, 0));
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        std::vector<std::vector<T>> matrix(size, std::vector<T>(size, 0));
```

File: matrix.h [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        for (int i = 0; i < size; i++) {
+            matrix[i][i] = 1;
+        }
+        return Matrix(matrix);
+    }
+
+    // Function to get the cofactor matrix (minor matrix)
+    Matrix getCofactor(size_t delrow, size_t delcol) {
+        size_t i = 0, j = 0;
+        std::vector<std::vector<T>> temp(size.first-1, std::vector<int>(size.second-1));
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        std::vector<std::vector<T>> temp(size.first-1, std::vector<T>(size.second-1)); 
```

File: matrix.h [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        }
+        return Matrix(matrix);
+    }
+
+    // Function to get the cofactor matrix (minor matrix)
+    Matrix getCofactor(size_t delrow, size_t delcol) {
+        size_t i = 0, j = 0;
+        std::vector<std::vector<T>> temp(size.first-1, std::vector<int>(size.second-1)); 
+        for (size_t row = 0; row < size.first; row++) {
+            for (size_t col = 0; col < size.second; col++) {
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        for (size_t row = 0; row < size.first; ++row) {
            for (size_t col = 0; col < size.second; ++col) {
```

File: matrix.h [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                if (row != delrow && col != delcol) {
+                    temp[i][j++] = mat[row][col];
+                    if (j == size.second - 1) {
+                        j = 0;
+                        i++;
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
                        ++i;
```

File: matrix.h [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                    }
+                }
+            }
+        }
+        return std::move(temp);
+    }
+
+    long long getTrace() {
+        long long result = 0;
+        for (size_t i = 0; i < size.first; i++) {
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
        for (size_t i = 0; i < size.first; ++i) {
```

File: matrix.h [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    bool isSquareMatrix(){
+        return getSize().first == getSize().second;
+    }
+
+    /// Returns the size of the matrix, first is number of rows, second is number of columns.
+    std::pair<uint64_t, uint64_t> getSize() const { 
+        return {matrix.size(), matrix[0].size()};
+    }
```
Comment (dominikstrak2001) [inline matrix.h:78]:
```suggestion
    std::pair<uint64_t, uint64_t> getSize() const { 
        if(matrix.size() > 0) {
            return {matrix.size(), matrix[0].size()};
        } else {
            return {0, 0};
        }
    }
```

File: matrix.h [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+                if (j != getSize().second - 1) {
+                    outputStream << " ";
+                }
+            }
+            outputStream << "\n";
+        }
+    }
+
+    /// Checks if the vector of vectors is a valid matrix.
+    static bool vectorIsValidMatrix(std::vector<std::vector<T>> mat){
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
    static bool isValidMatrix(std::vector<std::vector<T>> &mat) {
```

**Review by dominikstrak2001 — APPROVED** [review]

File: matrix.h [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+        return getSize().first == getSize().second;
+    }
+
+    /// Returns the size of the matrix, first is number of rows, second is number of columns.
+    std::pair<uint64_t, uint64_t> getSize() const { 
+        return {matrix.size(), matrix[0].size()};
+    }
+
+    bool operator==(const Matrix &other) const {
+        if (size != other.size) {
```
Comment (Arapak) [inline matrix.h]: Shouldn't this be getSize() == other.getSize() ?

**Review by Arapak — COMMENTED** [review]
Maybe document Matrix class in README.md

File: utils.h [hunk 13]
```diff
@@ -46,4 +46,16 @@ void setupDirectories() {
     }
 }
 
+template <typename T>
+concept ConvertibleToInt64_t = std::convertible_to<T, int64_t>;
+
+template<ConvertibleToInt64_t T>
+int64_t changeVectorToInt64_t(std::vector<T> v) {
```
Comment (dominikstrak2001) [inline utils.h]:
```suggestion
int64_t changeVectorToInt64_t(std::vector<T> &v) {
```

File: matrix.h [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+                if (j != getSize().second - 1) {
+                    outputStream << " ";
+                }
+            }
+            outputStream << "\n";
+        }
+    }
+
+    /// Checks if the vector of vectors is a valid matrix.
+    static bool vectorIsValidMatrix(std::vector<std::vector<T>> &mat) {
```
Comment (dominikstrak2001) [inline matrix.h]:
```suggestion
    static bool vectorIsValidMatrix(const std::vector<std::vector<T>> &mat) {
```

**Pull request merged**